### PR TITLE
airbyte-cron: update connector definitions from remote

### PR DESCRIPTION
**Pull request by pedroslopez**
## What

closes https://github.com/airbytehq/airbyte-cloud/issues/2542

Adds functionality to update source and destination definitions from the remote catalog at runtime, on an interval. 

This builds on the remote definitions provider introduced in #16018 

## How

- Set up airbyte-cron to connect to the configs db by following in airbyte-workers footsteps 
- Add a new DefinitionsUpdater task for airbyte-cron scheduled to run every 30 secs
- Move definitions update logic to new helper ApplyDefinitionsHelper to re-use between bootloader/cron and standardize between oss/cloud. This will replace cloud's ApplySeedsToDefinitions class.
- Use deployment mode env var to determine if we should overwrite everything or be more granular in how the update is performed
- Allow using `UPDATE_DEFINITIONS_CRON_ENABLED` env var to turn on / off these updates

Also fixed a couple of things while in here:
- When updating definitions for OSS, fix crash that occurs if there's a non-semver docker image tag in the db

## Recommended reading order
1. `ApplyDefinitionsHelper`
2. `DefinitionsUpdater`

## 🚨 User Impact 🚨
This will be off by default on OSS, so no impact.

### CODE REVIEW

File: airbyte-cron/src/main/java/io/airbyte/cron/config/DatabaseBeanFactory.java
Comment (pedroslopez) [reply airbyte-cron/src/main/java/io/airbyte/cron/config/DatabaseBeanFactory.java:31]: cc @jdpgrailsdev

Comment (jdpgrailsdev) [reply airbyte-cron/src/main/java/io/airbyte/cron/config/DatabaseBeanFactory.java:31]: @pedroslopez Yes...if we move any class that is annotated with Micronaut annotations to a library AND include the proper dependencies so that the annotation processor can find those classes during build time, the class will be automatically picked up by Micronaut at runtime from a library.  I have done this in my airbyte-workers PR with the :airbyte-config:config-models module.  I would also recommend doing this for the `ApplyDefintionsHelper`, which can just have the annotations placed directly on it (e.g. `@Singleton`).

**Comment by jdpgrailsdev** [comment]
@pedroslopez LGTM.  We should include @benmoriceau on this review as well, as he is working on a PR for `airbyte-cron` that may conflict with this one.

File: airbyte-cron/src/main/java/io/airbyte/cron/config/DatabaseBeanFactory.java
Comment (benmoriceau) [inline airbyte-cron/src/main/java/io/airbyte/cron/config/DatabaseBeanFactory.java:39]: @jdpgrailsdev Won't this be conflicting with the Bean declaration that is added in your review? It will complains that multiple Beans with this type are present.

Comment (jdpgrailsdev) [reply airbyte-cron/src/main/java/io/airbyte/cron/config/DatabaseBeanFactory.java:39]: @benmoriceau I'm not sure how Micronaut handles that.  It may just replace it.  We are going to have to work out the duplication once everything lands, most likely by extracting it to a shared location.

Comment (pedroslopez) [reply airbyte-cron/src/main/java/io/airbyte/cron/config/DatabaseBeanFactory.java:39]: Just to be clear here: is the recommendation then that it's ok to merge as-is and de-dupe when both are merged? If so I would love a general review / approval from either of you on the platform team 😄

File: airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java
Comment (davinchia) [inline airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1920]: for my understanding, why does this have to change to load configs from the remote?

File: airbyte-cron/src/main/java/io/airbyte/cron/DefinitionsUpdater.java
Comment (davinchia) [inline airbyte-cron/src/main/java/io/airbyte/cron/DefinitionsUpdater.java:21]: Add a javadoc.

File: airbyte-cron/src/main/resources/application.yml
Comment (davinchia) [inline airbyte-cron/src/main/resources/application.yml:8]: Curious, why default to false?

File: docker-compose.yaml
Comment (davinchia) [inline docker-compose.yaml:197]: all of these should be sorted.

How do we want OSS users to interact with this? If this is turned on for them, we should also update .env file for users that run docker compose up and the kustomize file (since we are still working on Helm).

File: airbyte-cron/src/main/java/io/airbyte/cron/DefinitionsUpdater.java
Comment (davinchia) [inline airbyte-cron/src/main/java/io/airbyte/cron/DefinitionsUpdater.java:19]: I'm not as familiar with micronaut, is this because we want to enforce only one copy?

File: airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java
Comment (davinchia) [reply airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java:134]: to make sure I follow along:
- this block is reimplemented in Cloud, where the bootloader is wrapped (or maybe this hasn't been done yet)
- we intend to reuse the DefintionsUpdater class in the wrapped bootloader in Cloud.

is that right?

Want to make sure I understand the dependency graph of what is used where.

**Review by davinchia — APPROVED** [review]
I'm not a micronaut expert, so I focused my review on the updated logic.

Looks sane. Some recommendations on how to clear things up. Couple of questions for my understanding.

File: airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java
Comment (pedroslopez) [reply airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1920]: It doesn't - it's an unrelated error that has been around that I fixed while I was in here. Basically if you had anything in your database that didn't follow semver (e.g. a `dev` version of a connector) the bootloader would crash when updating definitions

File: docker-compose.yaml
Comment (pedroslopez) [reply docker-compose.yaml:197]: For now, I don't expect OSS users to use this - they _could_  enable it and get auto-updating connectors, but it's mainly meant for cloud right now since that was our main focus. We're currently also not publishing an OSS catalog. I think we want to do something slightly different for officially bringing auto-updates to OSS (see evan's doc [here](https://docs.google.com/document/d/1SxsauSI2_mYduhLtHldWfjEGXIWtIiubwQInE0N0fcE/edit#)). I just added this in OSS because eventually the two upgrade mechanisms between cloud/oss will converge.

File: airbyte-cron/src/main/resources/application.yml
Comment (pedroslopez) [reply airbyte-cron/src/main/resources/application.yml:8]: Just because we don't want to enable this in OSS yet - more info in my other comment

File: airbyte-cron/src/main/java/io/airbyte/cron/DefinitionsUpdater.java
Comment (pedroslopez) [reply airbyte-cron/src/main/java/io/airbyte/cron/DefinitionsUpdater.java:19]: I'm also not super familiar with micronaut - I just copied this from the other tasks that had been defined but yes, that was my assumption. Micronaut takes care of instantiating it once and passing the dependencies.

File: airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java
Comment (pedroslopez) [reply airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java:134]: @davinchia yep! Cloud replaces this with this method here https://github.com/airbytehq/airbyte-cloud/blob/25d80a68b8f55499c2b4c0094d1956be18943e79/airbyte-bootloader-wrapped/src/main/java/io/airbyte/bootloader/wrapped/BootloaderAppWrapped.java#L129-L141

Which is currently using cloud's own [ApplySeedToDefinitionsClass](https://github.com/airbytehq/airbyte-cloud/blob/25d80a68b8f55499c2b4c0094d1956be18943e79/airbyte-bootloader-wrapped/src/main/java/io/airbyte/bootloader/wrapped/ApplySeedToDefinitions.java). After this is in, that ApplySeedToDefinitions class will be removed and we'll instead use our new shared DefinitionsUpdater

Comment (davinchia) [reply airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java:134]: Nice. I remember a duplicate class in Cloud and wanted to make sure we clean it up. Great to hear we are.

File: airbyte-config/init/src/main/java/io/airbyte/config/init/ApplyDefinitionsHelper.java
Comment (davinchia) [reply airbyte-config/init/src/main/java/io/airbyte/config/init/ApplyDefinitionsHelper.java:54]: If someone isn't going to pick this up in the next few weeks, we should create an issue and link it here so we don't forget.

File: airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java
Comment (davinchia) [reply airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/DatabaseConfigPersistence.java:1920]: got it. I think you pointed it out in your PR description - will also be nice to leave a comment here for future PRs so it's easy to associate :)

**Comment by pedroslopez** [comment]
Hi @gosusnp / @jdpgrailsdev could I get confirmation that this is ok to merge via a review/approval? I have merged in the latest changes from master that remove the airbyte-worker dependency and my changes here are not affected. 

Note that I'm defining my own `DatabaseBeanFactory` here which is why I don't need to depend on `airbyte-worker`, but I think de-duplicating this can be tackled as part of the micronaut tech debt epic that was being discussed.

**Comment by jdpgrailsdev** [comment]
@pedroslopez I'm fine with merging this.  We will want to do some cleanup later to avoid duplicating the configuration factory bits, but that shouldn't hold this up if we have confirmed that the `airbyte-workers` bit has been removed.  I'll defer to @benmoriceau and @gosusnp to make sure that there isn't something in here that would impact the work they are doing to extract some classes from `airbyte-workers` for use in this service (I don't believe that this is an issue, but worth double checking).

**Review by gosusnp — APPROVED** [review]
This introduces a new path for reading protocol versions which is missing some recent changes. I put the fix up in https://github.com/airbytehq/airbyte/pull/16796

**Pull request merged**
